### PR TITLE
[CUDA] Add pattern to collapse vector.transfer_read to 2D when possible

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
This is needed to lower these ops in the tensorcore pipeline

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
High level comment before I go into details here. Do we want to use reshapes or rank reducing subviews. Reshapes aren't the preferred way to handle this in general. When  something similar was needed on the CPU side rank reducing subviews worked

**Comment by nirvedhmeshram** [comment]
Maybe I am missing something here, but I did not use any reshapes here, basically collapsed all dims from 0 to N-1 to a single dim so that the 2D memref we had is made of the newly collapsed dim and the innermost (rightmost) dim. This was only done for cases where there is only one non unit-dim in the dims to be collapsed. Otherwise we bail.

**Comment by MaheshRavishankar** [comment]
> > High level comment before I go into details here. Do we want to use reshapes or rank reducing subviews. Reshapes aren't the preferred way to handle this in general. When something similar was needed on the CPU side rank reducing subviews worked
> 
> Maybe I am missing something here, but I did not use any reshapes here, basically collapsed all dims from 0 to N-1 to a single dim so that the 2D memref we had is made of the newly collapsed dim and the innermost (rightmost) dim. This was only done for cases where there is only one non unit-dim in the vector shape of the dims to be collapsed. Otherwise we bail.
> 

I was referring to collapsing dimensions as the reshape. You can take a subview of the original memref with all the dimensions you don't need reduced away. Reshapes don't compose. Here it works cause you have the full memref. If you had a memref that is a subview it would fail. So reshapes are general avoided and rank reducing subviews are preferred

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?

**Comment by ThomasRaoux** [comment]
> I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> 
> I am not able to figure out the context from the lit tests added.

This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.

If you look at the test it has the IR we are looking at:
`  %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`

here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 1]
```diff
@@ -91,21 +91,131 @@ static void createAsyncGroups(FuncOp funcOp) {
   }
 }
 
+static Value collapseMemRefTo2D(PatternRewriter& rewriter, mlir::Location loc,
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:94]: Can you add some basic documentation for this function and the one below.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 2]
```diff
@@ -91,21 +91,131 @@ static void createAsyncGroups(FuncOp funcOp) {
...
+  SmallVector<int64_t, 4> strides;
+  if (failed(getStridesAndOffset(t, strides, offset))) return AffineMap();
+  strides.pop_back();
+  int64_t firstStride = strides.back();
+  for (int64_t& i : strides) i = i / firstStride;
+  return makeStridedLinearLayoutMap(strides, offset, t.getContext());
+}
 namespace {
 
+struct FlattenTransferReadOp : public OpRewritePattern<vector::TransferReadOp> {
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:116]: I would also add some doc here explaining what the pattern is trying to do.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 3]
```diff
@@ -91,21 +91,131 @@ static void createAsyncGroups(FuncOp funcOp) {
...
+        vectorType.getRank() - 2;  // default index for the outermost dim of
+                                   // collapsed vector in the orginal vector
+    for (int i = 0; i < vectorType.getRank() - 1; i++) {
+      if (vectorType.getShape()[i] != 1) {
+        numberOfNonUnitDims++;
+        indexOfOuterMostUnitDim = i;
+      }
+    }
+    // Bail out if 2D vector cannot be formed
+    if (numberOfNonUnitDims > 2) {
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:142]: nit: should be `numberOfNonUnitDims != 2`?

**Review by ThomasRaoux — APPROVED** [review]
looks good to me once the couple missing comments are added

**Comment by MaheshRavishankar** [comment]
> > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > I am not able to figure out the context from the lit tests added.
> 
> This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> 
> If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> 
> here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.

isnt this same as 

```
%0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
%1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
```

IIUC this is what you want to achieve.

**Comment by ThomasRaoux** [comment]
> > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > I am not able to figure out the context from the lit tests added.
> > 
> > 
> > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> 
> isnt this same as
> 
> ```
> %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> ```
> 
> IIUC this is what you want to achieve.

I didn't realize subview allows to remove a unit dim in the middle. Is that legal?
Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.

**Comment by MaheshRavishankar** [comment]
> > > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > > I am not able to figure out the context from the lit tests added.
> > > 
> > > 
> > > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> > 
> > 
> > isnt this same as
> > ```
> > %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> > %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > IIUC this is what you want to achieve.
> 
> I didn't realize subview allows to remove a unit dim in the middle. Is that legal? Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.

Yup it is legal. There are some land mines hidden there, but they are less "lethal" than the land mins hidden under `memref.collapse/expand_shape` . It would be better to extend `vector` to mma to allow this.

**Comment by ThomasRaoux** [comment]
> 



> > > > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > > > I am not able to figure out the context from the lit tests added.
> > > > 
> > > > 
> > > > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > > > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > > > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> > > 
> > > 
> > > isnt this same as
> > > ```
> > > %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> > > %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > IIUC this is what you want to achieve.
> > 
> > 
> > I didn't realize subview allows to remove a unit dim in the middle. Is that legal? Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.
> 
> Yup it is legal. There are some land mines hidden there, but they are less "lethal" than the land mins hidden under `memref.collapse/expand_shape` . It would be better to extend `vector` to mma to allow this.

ok, I'm pretty scared of all the memref affine map part but we can try to relax some of the mma ops

**Comment by ThomasRaoux** [comment]
> > 
> 
> > > > > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > > > > I am not able to figure out the context from the lit tests added.
> > > > > 
> > > > > 
> > > > > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > > > > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > > > > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> > > > 
> > > > 
> > > > isnt this same as
> > > > ```
> > > > %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> > > > %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > IIUC this is what you want to achieve.
> > > 
> > > 
> > > I didn't realize subview allows to remove a unit dim in the middle. Is that legal? Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.
> > 
> > 
> > Yup it is legal. There are some land mines hidden there, but they are less "lethal" than the land mins hidden under `memref.collapse/expand_shape` . It would be better to extend `vector` to mma to allow this.
> 
> ok, I'm pretty scared of all the memref affine map part but we can try to relax some of the mma ops

The challenge with the memref layout is that we need to prove that they are contiguous and this keeps breaking, I have seen the same problem with load/store in conversion to llvm.
What's the problem with reshape ops again? I thought on the cpu side the subviews were used to avoid vector shape_cast?

**Comment by MaheshRavishankar** [comment]
> > > 
> > 
> > 
> > > > > > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > > > > > I am not able to figure out the context from the lit tests added.
> > > > > > 
> > > > > > 
> > > > > > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > > > > > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > > > > > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> > > > > 
> > > > > 
> > > > > isnt this same as
> > > > > ```
> > > > > %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> > > > > %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > IIUC this is what you want to achieve.
> > > > 
> > > > 
> > > > I didn't realize subview allows to remove a unit dim in the middle. Is that legal? Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.
> > > 
> > > 
> > > Yup it is legal. There are some land mines hidden there, but they are less "lethal" than the land mins hidden under `memref.collapse/expand_shape` . It would be better to extend `vector` to mma to allow this.
> > 
> > 
> > ok, I'm pretty scared of all the memref affine map part but we can try to relax some of the mma ops
> 
> The challenge with the memref layout is that we need to prove that they are contiguous and this keeps breaking, I have seen the same problem with load/store in conversion to llvm. What's the problem with reshape ops again? I thought on the cpu side the subviews were used to avoid vector shape_cast?

Its the same issue with the collapse. You need to prove that the collapse is contiguous and that keeps breaking as well. 

Even in the original `vector.transfer_read` here
```
%0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>
```
you are taking a `vector<16x1x8xf32>` out of a `memref<128x16x256xf32>` at an offset of `[2, 3, 4]` . I am not sure I follow how that is reading a contiguous memory out of the memref with such an instruction.

**Comment by ThomasRaoux** [comment]
> > > > 
> > > 
> > > 
> > > > > > > > I am missing the larger context as to what triggered these changes. Just raising here cause a similar issue on the CPU side was addressed using rank-reducing subviews. `memref.collapse_shape` are very hairy constructs. They require your original memref to be linear in memory for the reshape to be valid. In general that doesnt compose. So I'd like to get some idea of what the IR patterns that required this changes were?
> > > > > > > > I am not able to figure out the context from the lit tests added.
> > > > > > > 
> > > > > > > 
> > > > > > > This transformation is inspired from what was done for the cpu however the problem here is different and we cannot use a subview. The idea here is to transform the transfer_read into a 2D read than can be converted into an mma load. We need to be able to access the full span of the original memref so we cannot just take a subview. We want to collapse the higher dimensions instead.
> > > > > > > If you look at the test it has the IR we are looking at: ` %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>`
> > > > > > > here we want to be able to access data in the full `memref<128x16x256xf32>` but we want the target vector to be a 2D vector. To do that we collapse the higher dimension.
> > > > > > 
> > > > > > 
> > > > > > isnt this same as
> > > > > > ```
> > > > > > %0 = memref.subview %a[%c2, %c3, %c4] [16, 1, 8] [1, 1, 1] : memref<128x16x256xf32> to memref<16x8xf32, #map>
> > > > > > %1 = vector.transfer_read %0[%c0, %c0, %c0], %cst {...} : memref<16x8xf32, #map>, vector<16x8xf32>
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > IIUC this is what you want to achieve.
> > > > > 
> > > > > 
> > > > > I didn't realize subview allows to remove a unit dim in the middle. Is that legal? Right now I don't think the vector to mma will support this kind of map in the memref but it could be extended.
> > > > 
> > > > 
> > > > Yup it is legal. There are some land mines hidden there, but they are less "lethal" than the land mins hidden under `memref.collapse/expand_shape` . It would be better to extend `vector` to mma to allow this.
> > > 
> > > 
> > > ok, I'm pretty scared of all the memref affine map part but we can try to relax some of the mma ops
> > 
> > 
> > The challenge with the memref layout is that we need to prove that they are contiguous and this keeps breaking, I have seen the same problem with load/store in conversion to llvm. What's the problem with reshape ops again? I thought on the cpu side the subviews were used to avoid vector shape_cast?
> 
> Its the same issue with the collapse. You need to prove that the collapse is contiguous and that keeps breaking as well.
> 
> Even in the original `vector.transfer_read` here
> 
> ```
> %0 = vector.transfer_read %a[%c2, %c3, %c4], %cst {in_bounds = [true, true, true]} : memref<128x16x256xf32>, vector<16x1x8xf32>
> ```
> 
> you are taking a `vector<16x1x8xf32>` out of a `memref<128x16x256xf32>` at an offset of `[2, 3, 4]` . I am not sure I follow how that is reading a contiguous memory out of the memref with such an instruction.

Right the whole memref doesn't need to be contiguous, basically it needs to be able contiguous along the most inner dimension and have a constant stride to the second dimension.

Looking at it, it might not be too bad to support, I sent a patch that should allow us to handle this case upstream I think:
https://reviews.llvm.org/D122452

I think we can land this patch with subview in the iree in the meantime and things will connect when we integrate.

**Comment by MaheshRavishankar** [comment]
> I think we can land this patch with subview in the iree in the meantime and things will connect when we integrate.

You mean reimplement this patch with subview? It currently isnt using subview.

**Comment by ThomasRaoux** [comment]
> > I think we can land this patch with subview in the iree in the meantime and things will connect when we integrate.
> 
> You mean reimplement this patch with subview? It currently isnt using subview.

Yes I mean change it to use subview and land it. (meaning we don't need to wait for the upstream change).

If someone can review the other patch in the meantime this would be nice :)

**Comment by MaheshRavishankar** [comment]
> > > I think we can land this patch with subview in the iree in the meantime and things will connect when we integrate.
> > 
> > 
> > You mean reimplement this patch with subview? It currently isnt using subview.
> 
> Yes I mean change it to use subview and land it. (meaning we don't need to wait for the upstream change).
> 
> If someone can review the other patch in the meantime this would be nice :)

Stamped upstream review. 
Just for my information, is it because this e2e case is not tested on CI ? (If so probably need to test it on CI :P )

**Comment by ThomasRaoux** [comment]
> Stamped upstream review.
Thanks!

> Just for my information, is it because this e2e case is not tested on CI ? (If so probably need to test it on CI :P )
Well this case is not enabled (and not working anyway) until we turn it on in the config so it doesn't matter. This wouldn't break any execution test.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp (hunk 3 above)
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:233]: I wanted to support the case like vector<1x1x1x1> to vector <1x1> see the last lit test. But let me know if its better to bail out in a case like this and I can change it.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 4]
```diff
@@ -91,21 +91,221 @@ static void createAsyncGroups(FuncOp funcOp) {
...
+/// 3) The sizes are same as that of the src.
+/// For (3) when the shape is dynamic if the `src` is defined using an operation
+/// that implements the `ShapeAwareOpInterface` (like
+/// `hal.interface.binding.subspan`) then we can use that to check dynamic
+/// equality.
+/// Note: This could be written as a canonicalizer, but the subview formed
+/// when there are dynamic shapes involved will have affine maps
+/// that shouldnt be there. Resolving that is a pain. So dont generate the
+/// subview to begin with.
+static bool generatesNoOpSubView(Value src, ArrayRef<OpFoldResult> offsets,
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:107]: I thought that this code was familiar :) . If I am not mistaken this is taken from `LinalgBufferize`.
Some commentary, I hated writing this in the first place. This should really be a separate pattern that kicks in with canonicalize with IREE, but dont know how to do that.
Could you move this into a utility function in `Codegen/Utils`

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 5]
```diff
@@ -91,21 +91,221 @@ static void createAsyncGroups(FuncOp funcOp) {
...
+    Value vector = transferReadOp.vector();
+    VectorType vectorType = vector.getType().cast<VectorType>();
+    Value source = transferReadOp.source();
+    MemRefType sourceType = source.getType().dyn_cast<MemRefType>();
+    // Contiguity check is valid on tensors only.
+    if (!sourceType) return failure();
+    // Already 2D or lower nothing to do.
+    if (vectorType.getRank() < 3) return failure();
+    int numberOfNonUnitDims = 1;  // The innermost dim is always considered non
+                                  // unit as it wont be collapsed
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:222]: Nit: `s/wont be collapsed/wont be dropped`

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 6]
```diff
@@ -91,21 +91,221 @@ static void createAsyncGroups(FuncOp funcOp) {
...
+    int indexOfOuterMostUnitDim =
+        vectorType.getRank() - 2;  // default index for the outermost dim of
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:224]: I dont know the details, but the comment is not clear to me.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks for addressing my concerns!

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:107]: we actually don't need this code, we only care about the rank reducing case so we can skip the code calling it below

Comment (nirvedhmeshram) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:107]: Makes sense, in that case I wont move the pattern to utils

**Review by ThomasRaoux — COMMENTED** [review]
a big chunk of code can be removed other than that it looks good

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp (hunk 6 above)
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp]: I added a detailed description of this variable, hopefully it makes things clear. The name of the variable should have been and is changed to `indexOfOuterNonUnitDim`

**Comment by ThomasRaoux** [comment]
LGTM!

**Pull request merged**